src/common/Gateway.sol

```diff
@@ -267,58 +275,75 @@ contract Gateway is Auth, IGateway, Recoverable {
 
         IAdapter[] memory adapters_ = adapters[centrifugeId];
         require(adapters[centrifugeId].length != 0, EmptyAdapterSet());
 
-        uint64 batchGasLimit_ =
+        uint128 batchGasLimit_ =
             (isBatching) ? batchGasLimit[centrifugeId][poolId] : gasService.gasLimit(centrifugeId, batch);
-        uint64 proofGasLimit = gasService.gasLimit(centrifugeId, proof);
 
         for (uint256 i; i < adapters_.length; i++) {
-            IAdapter currentAdapter = IAdapter(adapters_[i]);
             bool isPrimaryAdapter = i == PRIMARY_ADAPTER_ID - 1;
-            bytes memory payload = isPrimaryAdapter ? batch : proof;
-            uint64 gasLimit = isPrimaryAdapter ? batchGasLimit_ : proofGasLimit;
-
-            uint256 consumed = currentAdapter.estimate(centrifugeId, payload, gasLimit);
+            uint256 consumed = adapters_[i].estimate(centrifugeId, isPrimaryAdapter ? batch : proof, batchGasLimit_);
 
             if (transactionPayer != address(0)) {
                 require(consumed <= fuel, NotEnoughTransactionGas());
                 fuel -= consumed;
             } else {
-                if (consumed <= subsidy[poolId]) {
-                    subsidy[poolId] -= consumed;
+                if (consumed <= subsidy[poolId].value) {
+                    subsidy[poolId].value -= uint96(consumed);
                 } else {
                     consumed = 0;
                 }
             }
 
-            currentAdapter.send{value: consumed}(
+            adapters_[i].send{value: consumed}(
                 centrifugeId,
-                payload,
-                gasLimit,
-                transactionPayer != address(0) ? transactionPayer : address(this)
+                isPrimaryAdapter ? batch : proof,
+                batchGasLimit_,
+                transactionPayer != address(0) ? transactionPayer : subsidy[poolId].refund
             );
 
             bytes32 batchId = keccak256(abi.encodePacked(localCentrifugeId, centrifugeId, batch));
             if (isPrimaryAdapter) {
-                emit SendBatch(centrifugeId, batchId, batch, currentAdapter);
+                emit SendBatch(centrifugeId, batchId, batch, adapters_[i]);
             } else {
-                emit SendProof(centrifugeId, batchId, proof, currentAdapter);
+                emit SendProof(centrifugeId, batchId, proof, adapters_[i]);
             }
         }
     }
 
-    function subsidizePool(PoolId poolId) external payable {
-        subsidy[poolId] += msg.value;
+    function _closeTransaction() internal {
+        if (transactionPayer == address(0)) return;
+
+        if (fuel > 0) {
+            (bool success,) = transactionPayer.call{value: fuel}(new bytes(0));
+
+            if (!success) {
+                // If refund fails, move remaining fuel to global pot
+                subsidy[PoolId.wrap(0)].value += uint96(fuel);
+                emit SubsidizePool(PoolId.wrap(0), address(0), fuel);
```

Comment:
```suggestion
                emit SubsidizePool(PoolId.wrap(0), gateway, fuel);
```
Here sender should not be the `Gateway`? Or even better, the `transactionPayer`?

Reply:
`gateway` makes sense! Will be `address(this)` then

I don't really like `transactionPayer since I feel it should be separated between direct subsidies from a user and a repayment after a tx.